src/Adapter/Order/CommandHandler/ChangeOrderDeliveryAddressHandler.php

```diff
@@ -77,14 +99,91 @@ public function handle(ChangeOrderDeliveryAddressCommand $command)
+            $cart->setDeliveryOption([
+                (int) $cart->id_address_delivery => $this->formatLegacyDeliveryOptionFromCarrierId($order->id_carrier),
+            ]);
+            $cart->update();
+
+            // gift could have been added/deleted when changing delivery address
+            $this->synchronizeOrderWithCart($order, $cart, $comparator);
+
+            $order->id_address_delivery = $address->id;
+            $this->orderDetailTaxUpdater->updateOrderDetailsTaxes($order);
+            $this->orderAmountUpdater->update($order, $cart);
+        } finally {
+            $this->contextStateManager->restorePreviousContext();
+        }
+    }
+
+    /**
+     * @param Order $order
+     * @param Cart $cart
+     * @param CartProductsComparator $productsComparator
+     */
+    private function synchronizeOrderWithCart(
+        Order $order,
+        Cart $cart,
+        CartProductsComparator $productsComparator
+    ): void {
+        $modified = $productsComparator->getModifiedProducts();
+        foreach ($modified as $productUpdate) {
+            $orderDetail = $this->getOrderDetail($productUpdate, $order, $cart);
+            if (null === $orderDetail) {
+                continue;
+            }
+            $quantity = $productUpdate->isCreated()
+                ? $productUpdate->getDeltaQuantity()
+                : $orderDetail->product_quantity + $productUpdate->getDeltaQuantity();
+            $orderInvoice = $orderDetail->id_order_invoice != 0 ? new OrderInvoice($orderDetail->id_order_invoice) : null;
+
+            $this->orderProductQuantityUpdater->update(
+                $order,
+                $orderDetail,
+                $quantity,
+                $orderInvoice,
+                false
+            );
+        }
+    }
+
+    /**
+     * @param CartProductUpdate $productUpdate
+     * @param Order $order
+     * @param Cart $cart
+     *
+     * @return OrderDetail|null
+     */
+    private function getOrderDetail(CartProductUpdate $productUpdate, Order $order, Cart $cart): ?OrderDetail
+    {
+        foreach ($order->getProducts() as $product) {
+            $combinationId = $productUpdate->getCombinationId() ? $productUpdate->getCombinationId()->getValue() : 0;
+            if (
+                (int) $product['product_id'] === $productUpdate->getProductId()->getValue()
+                && (int) $product['product_attribute_id'] === $combinationId
+            ) {
+                return new OrderDetail($product['id_order_detail']);
+            }
+        }
+
+        foreach ($cart->getProducts() as $product) {
+            $combinationId = $productUpdate->getCombinationId() ? $productUpdate->getCombinationId()->getValue() : 0;
+            if (
+                (int) $product['id_product'] === $productUpdate->getProductId()->getValue()
+                && (int) $product['id_product_attribute'] === $combinationId
+            ) {
+                $orderDetail = new OrderDetail();
```

Comment:
Maybe I'm wrong. But here you create an orderDetail for the product without checking if it already exists